packages/contracts/contracts/SmartInvoiceSplitEscrow.sol

```diff
@@ -0,0 +1,587 @@
+        uint256 providerAmount = amount - daoAmount;
+        IERC20(token).safeTransfer(dao, daoAmount);
+        IERC20(token).safeTransfer(provider, providerAmount);
+        released = released + amount;
+        _remaining = remaining;
+    }
+
+    /**
+     * @dev External function to release funds from the contract to the provider.
+     * Uses the internal `_release` function to perform the actual release.
+     * @param _token The milestones to release funds to
+     */
+    function releaseTokens(address _token) external override nonReentrant {
+        if (_token == token) {
+            _release();
+        } else {
+            require(_msgSender() == client, "!client");
+            uint256 balance = IERC20(_token).balanceOf(address(this));
+            IERC20(_token).safeTransfer(provider, balance);
+        }
+    }
+
+    /**
+     * @dev Internal function to withdraw funds from the contract to the client.
+     */
+    function _withdraw() internal {
+        require(!locked, "locked");
+        require(block.timestamp > terminationTime, "!terminated");
+        uint256 balance = IERC20(token).balanceOf(address(this));
+        require(balance > 0, "balance is 0");
+
+        IERC20(token).safeTransfer(client, balance);
+        milestone = amounts.length;
+
+        emit Withdraw(balance);
+    }
+
+    /**
+     * @dev External function to withdraw funds from the contract to the client.
+     * Uses the internal `_withdraw` function to perform the actual withdrawal.
+     */
+    function withdraw() external override nonReentrant {
+        return _withdraw();
+    }
+
+    /**
+     * @dev External function to withdraw funds from the contract to the client.
+     * Uses the internal `_withdraw` function to perform the actual withdrawal.
+     * @param _token The token to withdraw
+     */
+    function withdrawTokens(address _token) external override nonReentrant {
+        if (_token == token) {
+            _withdraw();
+        } else {
+            require(block.timestamp > terminationTime, "!terminated");
+            uint256 balance = IERC20(_token).balanceOf(address(this));
+            require(balance > 0, "balance is 0");
+
+            IERC20(_token).safeTransfer(client, balance);
+        }
+    }
+
+    /**
+     * @dev External function to lock the contract.
+     * @param _details Details of the dispute
+     */
+    function lock(bytes32 _details) external payable override nonReentrant {
+        require(!locked, "locked");
+        uint256 balance = IERC20(token).balanceOf(address(this));
+        require(balance > 0, "balance is 0");
+        require(block.timestamp < terminationTime, "terminated");
+        require(_msgSender() == client || _msgSender() == provider, "!party");
+
+        if (resolverType == ADR.ARBITRATOR) {
+            disputeId = IArbitrator(resolver).createDispute{value: msg.value}(
+                NUM_RULING_OPTIONS,
+                abi.encodePacked(details)
```

Comment:
I know this is probably not in the scope of the added functionality, but should this be the `_details` function arg instead of the `details` state var?